### ISSUE TITLE
Allow for user tokens to be invalidated

### ISSUE THREAD

**Issue by stridger** [traccar/traccar#5273]
Currently user tokens are not stored in the DB or anywhere and thus cannot be invalidated. This presents a security issue if tokens get leaked and it was possible to invalidate them before 5.3. 

Ideally the system would know about what tokens exist and would allow to invalidate them.

My personal context is that we use Traccar for a school bus system where parents track the buses. So the token URLs are shared with all parents and updated/invalidated on a yearly basis. I can work with the new system for now, but if the tokens ever leaked I would probably have to hack the system to delete and recreate user accounts etc.

**Comment by tananaev** [comment]
I think the more modern approach to tokens is to have them as short lived instead of long lived tokens with an option to invalidate.

**Comment by stridger** [comment]
For tokens which are used in Software that is definitely true. I have however described my context above, where I simply want people to be able to login with a URL without a shared username/password etc. And for that it is important to be able to invalidate tokens when necessary.

> I think the more modern approach to tokens is to have them as short lived instead of long lived tokens with an option to invalidate.

For tokens which are used in Software that is definitely true. I have however described my context above, where I simply want people to be able to login with a URL without a shared username/password etc. And for that it is important to be able to invalidate tokens when necessary.

**Comment by tananaev** [comment]
How many times have you done it so far? What are the stats?

**Comment by stridger** [comment]
So far we have been on 5.2 precisely because I was waiting for the new token system to mature. We have needed to reset the tokens on at least 2 occasions in that, no because tokens leaked but because the audience changed due to unforeseen circumstances.

But I cannot hold off the upgrade any longer, so I am upgrading to 5.12, but can foresee problems with not being able to invalidate user tokens.

**Comment by tananaev** [comment]
Can you elaborate on the audience change? Wouldn't that just be new user accounts naturally? I don't think I fully understand the context. And 2 occasions in what period of time?

**Comment by stridger** [comment]
So the accounts I have are shared amongst many people (up to 60 or so) that monitor the buses, but they can only monitor the bus their kid is on. When for example bus routes have to change, which occasionally happens, then the audience changes and so I reissue the tokens so that people can only see the buses they should have a view of. Routes rarely change and this was 2 occasions within 3 years, but the general point is that not being able to invalidate long lasting tokens can be a problem.

**Comment by tananaev** [comment]
It sounds to me that you're not using the system as intended. You should have an account for each user and then you can just link/unlink devices as needed.

**Comment by stridger** [comment]
Well issuing and managing 300 user accounts is not part of my scope as a volunteer and also people would hate having to login just to see where the bus is each day :-)

The point is that the system is perfectly usable for my scenario and has been working perfectly for 4 years (thank you for creating a system that we could use!).

The wider point is that revoking user tokens seems to be something that ought to be possible.

**Comment by tananaev** [comment]
I agree in general, but to prioritize this we would need a stronger case for it. For now we can keep this open to gauge the interest.

**Comment by lightseeker34** [comment]
Of all the things that I'm looking for, this is definitely a huge priority for me also. Would love a token management feature in the future. Thanks.

**Comment by improximus** [comment]
Yes, absolutely. Token invalidation would be a very essential thing. Please do consider adding it.

**Comment by ivan-avalos** [comment]
This should be considered a high-priority security risk. Changing the password should immediately invalidate all sessions for the user in question. 7 days is still too much for an incident where an abuser has admin privileges, which is already happening in one of our deployments.

My suggestion is to encode the username and hashed password of the user in the `TokenData` object that is used to generate the session token (correct me if I'm wrong), and use them to validate the session. This would cause all users to be logged out when the login credentials are modified.

**Comment by tananaev** [comment]
Token revocation is now implemented.

**Comment by ivan-avalos** [comment]
I just checked the patches, and if I'm not mistaken, it seems that in order to revoke a token (i.e. add it to `tc_revoked_tokens`), the token itself has to be passed as an argument of the `POST /session/token/revoke` request, so:

1. What happens if the token for the abuser's session is unknown?
2. How does it work for cookie-based sessions?

Also, it does't make sense to revoke individual tokens, as user/password are enough to generate new ones. Changing the login credentials should be enough (both security- and UX-wise) to invalidate any tokens associated with that user.

**Comment by tananaev** [comment]
1. You can check audit logs to find issued tokens.
2. Cookie sessions should be short lived and there's no plan to do any revocations for them.

**Comment by ivan-avalos** [comment]
1. This fixes the issue in theory, but it's highly impractical for non-sysadmins.
2. What information is encoded in the session token? If it's not just a random number, could it be possible to encode the username and hashed password within, so whenever they change in the database, all session tokens are automatically revoked? This would not require any modification to the database.

What I'm trying to say is that, in any reasonably secure system, a password change is expected to log out any sessions immediately, otherwise we're completely reliant on how benevolent the abuser is to not mess around with the system before its session expires.

**Comment by tananaev** [comment]
It should be possible to invalidate all active sessions, but it's not the same thing as token revocation.

**Comment by ivan-avalos** [comment]
Regarding cookie short-livedness, I just observed that the `Set-Cookie` header defines no expiration date, should I open a separate ticket for this or is this the intended behavior? (Or am I looking in the wrong place?)

```http
HTTP/1.1 200 OK
Server: nginx/1.22.1
Date: Wed, 12 Nov 2025 18:51:29 GMT
Content-Type: application/json
Content-Length: 710
Connection: keep-alive
Set-Cookie: JSESSIONID=XXXXXXXXXXXX.node0; Path=/; HttpOnly
Expires: Thu, 01 Jan 1970 00:00:00 GMT
```

**Comment by tananaev** [comment]
You can configure it.
